### DISCUSSION TITLE
More about SMTP settings: OAuth2 support

### DISCUSSION THREAD

**Discussion by LSNicholls** [serenity-is/Serenity#6591] [Help] [answered]
Hi folks,

I see that MailKit has the ability to do OAuth2 (https://github.com/jstedfast/MailKit/blob/master/GMailOAuth2.md) .  FWIW the sample is about gmail, but I will be addressing Office 365.  But I can't see:
1. where I could put the relevant configuration information into the SMTP settings portion of the appsettings.json file, in spite of @VictorTomaili 's previous help on #5911. 
2. whether or where Serenity.Extensions' EmailSender would actually use this information.   From the MailKit sample it looks like you would need to use AuthorizeAsync, right?  And I don't see that anywhere, just client.Authenticate.

So, my questions:
* Has anybody succeeded in doing this ? Can you share?  I found some related stuff in the wiki for general authentication purposes, but nothing that applied to the email facility.
* Suppose nobody has, and I want to either override or extend EmailSender class.  **How would I tell Serenity to use my class instead of the built-in one?**  Sorry, this is probably a ridiculously obvious thing to some of you, but I don't know where to start.
 
FWIW the MailKit sample is about gmail, but I will be addressing Office 365.  The relevant docs for how this should be donefrom MS's POV are here https://learn.microsoft.com/en-us/exchange/client-developer/exchange-web-services/how-to-authenticate-an-ews-application-by-using-oauth and the Office 365 sample in MailKit docs is here http://www.mimekit.net/docs/html/T_MailKit_Security_SaslMechanismOAuth2.htm, so it *should* be possible with either server type, and either way it looks like AuthorizeAsynch method would have to be invoked.

Thanks to anybody for your thoughts.

**Comment by volkanceylan** [comment, marked as answer]
I don't think OAuth2 works properly with MailKit and a server session. I think i also tried that before and it did not work. And even if it did, we would not be supporting how to implement that as it is a Gmail and MailKit issue, not Serenity itself.

You should implement a class with IEmailSender interface and register it in Startup.cs, see .NET DI docs: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-7.0

**Comment by LSNicholls** [comment]
Thanks Volkan.  That makes sense.   I replace the   services.AddSingleton<Serenity.Extensions.IEmailSender, Serenity.Extensions.EmailSender>(); line in startup with my own class, and everything just works because the framework code says    [FromServices] IEmailSender emailSender.

This is all new to me (dependency injection as implemented in C#) although it philosophically matches how I have done pretty much everything in other environments. Basically, factory pattern. I can see that the Serenity framework implements it with some rigor and that makes everything possible on my end.  It's really fabulous.

Can I ask if you logged a bug with MailKit about this, and how long ago you tried?  Just in case I can shortcut my problem ;-).

Interesting that in the MailKit example for office 365 does not show use of a client secret although the MS documentation does.  I wonder if that is part of the server problem considering it should be application-only, not delegated ,permission when used on the server.  Just a thought.

**Comment by volkanceylan** [comment]
I tried enough. That is the reason we are using Gmail with password not oauth for serenity.is. lf they fixed something in last three months dont knowm

**Comment by LSNicholls** [comment]
Thanks Volkan, not asking you to do more, just wondering how long ago the "try" was!  Appreciate you.
